### ISSUE TITLE
[BotC] Map chipchap's multicurrency transfers in ocp as exchanges

### ISSUE THREAD

**Issue by XaviP** [FreedomCoop/valuenetwork#333]
"Map chipchap's multicurrency transfers in ocp as exchanges (transfer events with diferent units or resource types)" -> Feature request read in the pad of last assembly. Please, develop a bit the requeriments and how is thought to be done.

Some doubts at first sight:
- I think it refers to something similar to faircoin account in freedomcoop, but it's so different, as the faircoin transactions always comes from or goes to an agent in the ocp system, but in the case of chipchap transactions wallet, the transactions can be done outside the ocp scope, and always we'll have events with agent = None (as not known)
- Maybe this feature has sense only in the case of collective agents that wants to be transparent with their members, showing all transactions done with the collective account. But for individual agents, does it has sense??

**Comment by fosterlynn** [comment]
A couple of other thoughts on this one:

1. We would need to decide how to support multi-currency wallets in OCP.  This we think can be done with the existing resource parent structure, so not a big deal, but some UI changes would be needed.

2. Maybe we should consider the new Kamasi UI as an option to unify the exchange views.  Can we get exchange data directly from chipchap, or is it better to bring it through OCP anyhow?  If from chipchap, do they already have an api, and if so what does it look like?  

3. If we get data directly from chipchap maybe that helps with the problem of null agents on the events, if chipchap has that data? Or, if chipchap is a temporary solution anyhow, then we can think more about how to modify the OCP event model to handle what information we have even if there isn't an internal agent.

**Comment by XaviP** [comment]
> We would need to decide how to support multi-currency wallets in OCP. This we think can be done with the existing resource parent structure, so not a big deal, but some UI changes would be needed.

Currently, the access to chipchap api is attached to the agent, not as a resource:
https://github.com/FreedomCoop/valuenetwork/blob/master/multicurrency/models.py#L9

> If we get data directly from chipchap maybe that helps with the problem of null agents on the events, if chipchap has that data? Or, if chipchap is a temporary solution anyhow, then we can think more about how to modify the OCP event model to handle what information we have even if there isn't an internal agent.

Currently, in faircoin transactions, when a transaction comes from or goes to an agent that doesn't exist in the system, it's defined as agent = None

**Comment by fosterlynn** [comment]
Thanks @XaviP 

>Currently, the access to chipchap api is attached to the agent, not as a resource

Got it, then there must be only one chipchap wallet per agent.  So for that agent we could create/update a parent resource with a child for each currency in the chipchap wallet, on the OCP side.  (Just brainstorming.)

>Currently, in faircoin transactions, when a transaction comes from or goes to an agent that doesn't exist in the system, it's defined as agent = None

Then we could just do the same for chipchap transactions, and see if any of that is a problem for anyone.  I suppose we can get the external address from the other apps if we ever need it; or link to the transaction; etc.  But doesn't sound like a problem for now then.

**Comment by XaviP** [comment]
> for that agent we could create/update a parent resource with a child for each currency in the chipchap wallet

Can you develop this a little bit? Which objects must be created in each transaction? (exchange, transfer, event...) Is it something like when a faircoin transaction is read in the blockchain for including it in ocp? See here:
https://github.com/FreedomCoop/valuenetwork/blob/master/valuenetwork/valueaccounting/service.py#L268

Sincerilly, I get lost with abstractions. The best help could be some prototipe code.

**Comment by fosterlynn** [comment]
That code is a good start.  It doesn't seem to include doing anything with a resource though.  Maybe we don't keep the resource information (balance) for faircoin addresses in OCP?  Or maybe that code is elsewhere?

Here is how the exchange stuff fits together:

![chipchap-ocp](https://user-images.githubusercontent.com/3776081/29142085-5d29f4da-7d16-11e7-8036-ad6dce355c36.png)

An exchange ties together one or more transfers (in this case just one I assume).  (Theoretically you wouldn't need an exchange if there is only one transfer (nothing is being exchanged) - but OCP requires an exchange anyhow.)  

Each transfer has a give event and a receive event.*  [This is because each event affects one resource.  This wasn't our original model; and it has changed again in ValueFlows; but this is how it is in this version of OCP.  It is perhaps too complex, sorry about that.]  Like in the case of transfers of currency, the give event will decrease the resource (wallet/address/etc) of the giving agent, and the receive event will increase the resource (wallet/address/etc.) of the receiving agent. 

* There is both a give and receive event if the transfer is within the network (OCP instance).  If the transfer is to or from someone outside the network (OCP instance), then you would need only the give or the receive event and not both - just the one that affects the resource of the agent in the network.  (Theoretically, the other event exists, but resides outside the context of the instance, so is not recorded there.)

For the multi-currency wallets, can one transaction include more than one currency?  This will make a difference in how we represent it in OCP.

The exchange type and transfer type are probably pre-created in the data, and used in a hard-coded way, as it appears the faircoin transfers do.  Event types are "give" or "receive", that's it.

I'll try to do some prototype code, or at least some pseudo-code.

**Comment by bhaugen** [comment]
>  Maybe we don't keep the resource information (balance) for faircoin addresses in OCP? 

No, we don't, at least as of now. We look it up on the chain when we need it. This has caused problems when the lookup fails, but when that lookup fails, so will any transactions they want to send with their balance.

**Comment by fosterlynn** [comment]
>No, we don't, at least as of now. We look it up on the chain when we need it. This has caused problems when the lookup fails, but when that lookup fails, so will any transactions they want to send with their balance.

OK then I guess we can do the same with chipchap wallets then.  That saves some work.  And keeps us from having problems with the amounts being out of sync.  So seems good to me.

**Comment by fosterlynn** [comment]
>OK then I guess we can do the same with chipchap wallets then. 

I'm thinking about our earlier recommendation to use resource parents to tie together the chipchap wallet and each currency resource within the wallet.  Since we talk about not storing any resources on OCP related to this, I guess we don't need any resource parent thing since there are no resources stored - the UI can handle it, by bundling all the different currency amounts into one assumed chipchap wallet by agent if it is needed.  Or maybe just listing each individual currency amount is enough.

**Comment by fosterlynn** [comment]
>> Maybe we don't keep the resource information (balance) for faircoin addresses in OCP?

>No, we don't, at least as of now. We look it up on the chain when we need it

@bhaugen are you sure?  I see this method `create_faircoin_resource` in @XaviP code link from above.

**Comment by bhaugen** [comment]
I'm actually not sure. I know how it was, but things have changed.

**Comment by XaviP** [comment]
Currently, there's a resource for each faircoin account but it doesn't track the balance, it is asked to the wallet/blockchain each time the user visits the account or the events list.

**Comment by fosterlynn** [comment]
So, actually the code ref you gave @XaviP is pretty close to what is needed.  Here is an outline:
```
Create and save exchange
Create and save transfer
If sending agent is in OCP:
     Create and save "give" event
If receiving agent is in OCP:
     Create and save "receive" event
```

In addition, before any of this happens you'll need to set up an exchange type with a transfer type in OCP, similar to what was done for Faircoin transactions.

And also before any transactions happen, if we do BotC the same as Faircoin (and why not?), you'll want to create some code to:
1. Create an EconomicResource for the "parent", which references the person's chipchap wallet.  Maybe when the person signs up for BotC?
2. Create an EconomicResource as a "child" of the first one for each currency that will be used.  Maybe just create one for each currency supported?  Or maybe there is another way to know what currencies a person will use?  Not sure of the trigger for this one.

All of those resources will have 0 quantity.

Does this make sense?  More questions?

**Comment by fosterlynn** [comment]
>I think it refers to something similar to faircoin account in freedomcoop, but it's so different, as the faircoin transactions always comes from or goes to an agent in the ocp system, but in the case of chipchap transactions wallet, the transactions can be done outside the ocp scope, and always we'll have events with agent = None (as not known)

@XaviP is it true that some chipchap transactions have no agents inside of OCP?

**Comment by XaviP** [comment]
> is it true that some chipchap transactions have no agents inside of OCP?

Not really, always the sender or receiver will be the agent in ocp. But I mean that it's possible that an agent uses the wallet always in chipchap system and never in the agent context of ocp, so it never will be internal ocp transactions. That's when I don't see any sense of track the history inside ocp.

**Comment by XaviP** [comment]
I have a doubt in implementing chipchap transactions in ocp. The balances and transactions can be in many currencies: faircoin, bitcoin, euros, dollars, etc... So, do I need an EconomicResource for each currency linked with the MulticurrencyAuth object? And each transaction as EconomicEvent should "belong" to the EconomicResource of its currency?

**Comment by bhaugen** [comment]
> We would need to decide how to support multi-currency wallets in OCP. This we think can be done with the existing resource parent structure, so not a big deal, but some UI changes would be needed.

I'm rethinking this, and Lynn and I talked about on our walk this morning. I think it is different behavior from parent. It's more like a container, which I am now implementing for LearnDeep. 

And parent is a field on ResourceType, not Resource, while these will be relationships between Resources.

I would somewhat like to do what I am doing for LearnDeep, which is to add a container field to EconomicResource, like this:
```
    container = models.ForeignKey('self', blank=True, null=True,
        verbose_name=_('container'), related_name='contents')
```
And I added a container to the BEHAVIOR_CHOICES for EconomicResourceType. The multi-account behavior could live in  https://github.com/FreedomCoop/valuenetwork/blob/master/multicurrency/models.py 

Does that make sense? I am reluctant to add anything specific for multi-currency to valueaccounting.models. Keep it in its own app.

> If you take a look in second transaction, it has "exchange_EURtoFAC" method. So, two currencies in a transaction.
> how to map a transaction in two different currencies like the second sample I've linked before?

How does such a transaction affect the multi-currency accounts? Would it decrement the EUR account and increment the FAC account? Also, does ChipChap charge a fee for the currency transfer?

**Comment by XaviP** [comment]
But I think we need a compatible structure with faircoin account trnsactions. I mean, when ocp needs to read economic events, it has to read some kind of info generated in a similar way in faircoin app, and in multicurrency app. Don't you think so?

**Comment by bhaugen** [comment]
I don't know yet. Need answers to these questions:
> How does such a transaction affect the multi-currency accounts? Would it decrement the EUR account and increment the FAC account? Also, does ChipChap charge a fee for the currency transfer?

It could be more like a transfer between accounts than a single transaction in OCP terms, which would mean 2 OCP EconomIcEvents, one for each OCP account, with a different currency for each, with maybe another transaction to pay ChipChap if necessary, which would make it into a 3-event Exchange.

**Comment by XaviP** [comment]
If we are going to get the balance through the chipchap api, as in faircoin accounts, and the balance will not be stored in the EconomicResource, as in faircoin EconomicResources too, then it has not use to include in ocp these transactions with method "exchange_EURtoFAC" or any internal (from agent to same agent) transactions.

This will make simpler to map only the "real" transactions, and get the balance from chipchap api, as in faircoin app.

What do you think?

**Comment by fosterlynn** [comment]
I agree completely with not storing any balance in 2 places.  Better to go to the system of record when it is needed.

**Comment by bhaugen** [comment]
@XaviP just so I am sure I understand, are you concluding that the exchange_EURtoFAC transactions do not need to be included in OCP? 

That's ok with me, I think I understand your reasons. Checking to see if I understood what you wrote above correctly. 

And does that mean you have a pretty good idea about how to implement these multi-currency accounts now? Or do you have any unresolved issues?
